core/src/test/scala/unit/kafka/security/auth/SimpleAclAuthorizerTest.scala

```diff
@@ -0,0 +1,218 @@
+
+    assertTrue("superuser always has access, no matter what acls.", simpleAclAuthorizer.authorize(session1, Read, resource))
+    assertTrue("superuser always has access, no matter what acls.", simpleAclAuthorizer.authorize(session2, Read, resource))
+  }
+
+  @Test
+  def testNoAclFound(): Unit = {
+    assertFalse("when acls = [],  authorizer should fail close.", simpleAclAuthorizer.authorize(session, Read, resource))
+  }
+
+  @Test
+  def testNoAclFoundOverride(): Unit = {
+    val props = TestUtils.createBrokerConfig(1, zkConnect)
+    props.put(SimpleAclAuthorizer.AllowEveryoneIfNoAclIsFoundProp, "true")
+
+    val cfg = KafkaConfig.fromProps(props)
+    val testAuthoizer: SimpleAclAuthorizer = new SimpleAclAuthorizer
+    testAuthoizer.configure(cfg.originals)
+    assertTrue("when acls = null or [],  authorizer should fail open with allow.everyone = true.", testAuthoizer.authorize(session, Read, resource))
+  }
+
+  @Test
+  def testAclManagementAPIs(): Unit = {
+    val user1 = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, username)
+    val user2 = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, "bob")
+    val host1 = "host1"
+    val host2 = "host2"
+
+    val acl1 = new Acl(user1, Allow, host1, Read)
+    val acl2 = new Acl(user1, Allow, host1, Write)
+    val acl3 = new Acl(user2, Allow, host2, Read)
+    val acl4 = new Acl(user2, Allow, host2, Write)
+
+    var acls = Set[Acl](acl1, acl2, acl3, acl4)
+
+    addAclAndVerify(acls)
+
+    //test addAcl is additive
+    val acl5: Acl = new Acl(user2, Allow, WildCardHost, Read)
+    simpleAclAuthorizer.addAcls(Set[Acl](acl5), resource)
+
+    acls += acl5
+
+    TestUtils.waitUntilTrue(() => simpleAclAuthorizer.getAcls(resource) == acls, "changes not propagated in timeout period.")
+
+    //test get by principal name.
+    assertEquals(Set(acl1, acl2), simpleAclAuthorizer.getAcls(user1))
+
+    //test remove acl from existing acls.
+    simpleAclAuthorizer.removeAcls(Set(acl1, acl5), resource)
+    acls -= acl1
+    acls -= acl5
+
+    TestUtils.waitUntilTrue(() => simpleAclAuthorizer.getAcls(resource) == acls, "changes not propagated in timeout period.")
+
+    //test remove all acls for resource
+    simpleAclAuthorizer.removeAcls(resource)
+    TestUtils.waitUntilTrue(() => simpleAclAuthorizer.getAcls(resource) == Set.empty[Acl], "changes not propagated in timeout period.")
+  }
+
+  @Test
+  def testLoadCache(): Unit = {
+    val user1 = new KafkaPrincipal(KafkaPrincipal.USER_TYPE, username)
+    val host1 = "host1"
+
+    val acl1 = new Acl(user1, Allow, host1, Read)
+    val acl2 = new Acl(user1, Allow, host1, Write)
+    val acls = Set[Acl](acl1, acl2)
+    simpleAclAuthorizer.addAcls(acls, resource)
+
+    val authorizer = new SimpleAclAuthorizer
+    authorizer.configure(config.originals)
+    assertEquals(acls, simpleAclAuthorizer.getAcls(resource))
+  }
+
+  private def addAclAndVerify(acls: Set[Acl]) {
```

Comment:
Could we make this more general to sth like the following? Then we can remove all individual usage of waitUntilTrue().

// return the new acl set after applying the changes to the original acl.
changeAclAndVerify(originalAcl: Set[Acl], addedAcl: Set[Acl], removedAcl: Set[Acl]): Set[Acl]